lib/navigation/preprocess_encoders.cpp

```diff
@@ -1,56 +1,106 @@
+#include "consts.hpp"
 #include "preprocess_encoders.hpp"
 
+#include "core/types.hpp"
+
 namespace hyped::navigation {
 
-EncodersPreprocessor::EncodersPreprocessor()
+EncodersPreprocessor::EncodersPreprocessor(core::ILogger &logger)
+    : logger_(logger),
+      num_consecutive_outliers_per_encoder_({0, 0, 0, 0}),
+      are_encoders_reliable_({true, true, true, true}),
+      num_reliable_encoders_(core::kNumEncoders),
+      max_consecutive_outliers_(10)
+{
+}
+
+std::optional<core::EncoderData> EncodersPreprocessor::processData(
+  const core::EncoderData &raw_encoder_data)
 {
-  // TODOLater: implement
+  const auto encoder_data = sanitise(raw_encoder_data);
+  const auto sensors      = checkReliable();
+  if (sensors == SensorChecks::kUnacceptable) { return std::nullopt; }
+  return encoder_data;
 }
 
-core::EncoderData EncodersPreprocessor::processData(const core::EncoderData)
+std::optional<EncodersPreprocessor::Statistics> EncodersPreprocessor::getStatistics(
+  const core::EncoderData &encoder_data) const
 {
-  /*
-  TODOLater: implement
-  basic plan:
-  - call detectOutliers. Return of that function is return of this function
-  - call checkRelaible
-
-  */
-  return {0, 0, 0, 0};
+  if (num_reliable_encoders_ > core::kNumEncoders
+      || num_reliable_encoders_ < core::kNumEncoders - 1) {
```

Comment:
I don't think so, because kNumEncoders represents the number of encoder sensors that we have initially and num_reliable_encoders represents the number of encoder sensors that are working at the moment. As we go on the sensors can either keep working properly or stop working. That's why num_reliable_encoders is only decremented and can never be greater than kNumEncoders

Reply:
exactly, but you have a check to see if it is greater which is then redundant?

Reply:
It's just there to catch any bugs I believe, I'm pretty sure this was implemented by @mifrandir? The number of unreliable encoders is only ever reduced but we don't explicitly check for any other values elsewhere in the code so would lead to some dangerous behavior if this isn't caught.

Reply:
since that is completely undefined behavior I think it should be dealt with separately (like a different kFatal log to emphasize this) than just logging that its an unsuitable amount of encoders

Reply:
Fair point. @aryansharma19?